cql3/type_json.cc

```diff
@@ -71,15 +72,23 @@ static sstring quote_json_string(const sstring& value) {
     return oss.str();
 }
 
-static int64_t to_int64_t(const rjson::value& value) {
+
+template <typename T> static T to_int(const rjson::value& value) {
+    int64_t result;
+
     if (value.IsInt64()) {
-        return value.GetInt64();
+        result = value.GetInt64();
     } else if (value.IsInt()) {
-        return value.GetInt();
+        result = value.GetInt();
     } else if (value.IsUint()) {
-        return value.GetUint();
+        result = value.GetUint();
     } else if (value.IsUint64()) {
-        return value.GetUint64(); //NOTICE: large uint64_t values will get overflown
+        uint64_t u64_result = value.GetUint64();
+
+        if (std::numeric_limits<T>::min() > u64_result || u64_result > std::numeric_limits<T>::max()) {
+            throw marshal_exception(format("Value {} out of range", u64_result));
+        }
+        return u64_result;
```

Comment:
I think this check is incorrect. Reproducer (compiler explorer: https://godbolt.org/z/Evvzjj5zj):

```cpp
template <typename T> 
T parse(uint64_t u64_result) { 
    if (std::numeric_limits<T>::min() > u64_result || u64_result > std::numeric_limits<T>::max()) {
        throw std::runtime_error("Value out of range");
    }
    return u64_result;
}

int main() {
    std::cout << parse<int>(123) << std::endl; // Throws! It should NOT throw.
}
```

Granted this is true, a test should be written to catch this problem.

Reply:
Right, I think you shouldn't compare signed and unsigned values. See also https://www.modernescpp.com/index.php/safe-comparisons-of-integrals-with-c-20

As @avelanarius said, if this causes a bug in practice, and our test didn't catch it, we need another test that does.

Reply:
I didn't add the test because there is no way to test this. RapidJSON parse integer as `uint64_t` only when the value is in range [2^63; 2^64) and then the error doesn't occur.